compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/AIETarget.cpp

```diff
@@ -164,131 +206,193 @@ LogicalResult AIETargetBackend::serializeExecutable(
   if (failed(maybeWorkDir)) return failure();
   auto workDir = maybeWorkDir.value();
 
-  SmallString<128> inputMlirPath(workDir);
-  llvm::sys::path::append(inputMlirPath, basename + ".aiecc.mlir");
-
-  std::string errorMessage;
-  {
-    auto inputMlirOut = openOutputFile(inputMlirPath, &errorMessage);
-    if (!inputMlirOut) {
-      return moduleOp.emitOpError() << "Failed to write MLIR: " << errorMessage;
-    }
-    moduleOp.print(inputMlirOut->os(), OpPrintingFlags().useLocalScope());
-    inputMlirOut->keep();
-  }
-
-  SmallString<128> aie2xclbin(options.mlirAieInstallDir);
-  llvm::sys::path::append(aie2xclbin, "bin", "aie2xclbin");
-  SmallString<128> xclbinPath(workDir);
-  llvm::sys::path::append(xclbinPath, basename + ".xclbin");
-  SmallString<128> npuInstPath(workDir);
-  llvm::sys::path::append(npuInstPath, basename + ".npu.txt");
-
   // collect names of kernels as they need to be in kernels.json
   // generated by `aie2xclbin`
   SmallVector<std::string> entryPointNames;
+  // collect the aie.device ops that need to be passed to the `aie2xclbin` tool
+  SmallVector<xilinx::AIE::DeviceOp> deviceOps;
+  // Map to keep track of which ordinal number belongs to which entry point,
+  // typically the order is sequential but that is not gauranteed
+  std::map<std::string, uint64_t> entryPointOrdinals;
   for (auto exportOp : variantOp.getExportOps()) {
+    uint64_t ordinal = 0;
+    if (std::optional<APInt> optionalOrdinal = exportOp.getOrdinal()) {
+      ordinal = optionalOrdinal->getZExtValue();
+    } else {
+      // For executables with only one entry point, linking doesn't kick in at
+      // all. So the ordinal can be missing for this case.
+      if (!llvm::hasSingleElement(variantOp.getExportOps())) {
+        return exportOp.emitError() << "should have ordinal attribute";
+      }
+    }
+
+    StringRef exportOpName = exportOp.getSymName();
+    deviceOps.push_back(getDeviceOpFromEntryPoint(moduleOp, exportOpName));
+
     // The xclbin kernel name, appended with instance name suffix (`:MLIRAIEV1`,
     // 10 chars) is required by the xclbinutil to have a length smaller or equal
     // to 64 right now. To have some additional wiggle room for suffix changes,
-    // we use the 48 first characters for the kernel name. This is okay to do
-    // for now because we are only supporting single entry point.
-    entryPointNames.emplace_back(exportOp.getSymName().substr(0, 48));
+    // we use the 42 first characters for the kernel name. We suffix ordinal
+    // number to the name to gaurantee uniqueness within the executable.
+    std::string entryPointName = llvm::join_items(
+        "_", exportOpName.substr(0, 42), std::to_string(ordinal));
```

Comment:
nit: wouldn't it be better to check and fail that the name is less than 42 chars?

Reply:
Not sure I follow, do you mean if the user and/or IREE's naming scheme on top of that gives a large name we dont compile?

Reply:
I think I see what you meant, you are talking of the resulting name, not the initial name? I added a check that the resulting name is less than 50 characters which leaves enough wiggle room for the utility tools to suffix more things to it.